src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -1185,6 +1180,34 @@ void ImageReplayer<I>::remove_image_status(bool force, Context *on_finish)
   ctx->complete(0);
 }
 
+template <typename I>
+void ImageReplayer<I>::remove_image_status_remote(bool force, Context *on_finish)
+{
+  if (!m_in_flight_op_tracker.empty()) {
+    dout(15) << "waiting for in-flight operations to complete" << dendl;
```

Comment:
What do we need this "waiting for in-flight operations to complete" twice? Or rather 3 times, because we already have one in `handle_shut_down`.

I think we need just one but in a proper place. May be just move this "waiting" thing in `handle_shut_down` up before `remove_image_status`? I don't know, there might have been some reason why it was done after `image_deleter::ImageRemoveRequest`, but right now I don't see reasons. So this is what I would try first.

Reply:
Ah I forgot that this was done also right before the call to remove_mirror_image_status. This is strange because I saw the update_mirror_image_status here https://github.com/ceph/ceph/blob/df32fa6eff6b88949b6446055b0ada2829784cc1/src/tools/rbd_mirror/ImageReplayer.cc#L893 queue an update to the MirrorStatusUpdater in the work_queue and this update was done just between the try_remove_image_status from the MirrorStatusUpdate and the actual update_task when the actual removal of the status is done.

Let me check what's happening here...

Reply:
Ah buy wait in master the wait for in flight... is in the correct place but here it's done after the remove_image_status. So I just need to move some calls after that.

Reply:
Should be fine now!